go/state/state_db_test.go

```diff
@@ -4389,6 +4389,79 @@ func TestNonCommittableStateDB_resetState_ClearsTransientStorage(t *testing.T) {
 	}
 }
 
+func TestStateDB_ContractCanBeCreatedAndDeletedInTheSameTransaction(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	// the account should exist prior to CreateContract call
+	mock.EXPECT().Exists(address1).Return(true, nil)
+	db.CreateContract(address1)
+
+	// suicide the newly created contract in the same tx scope
+	db.SuicideNewContract(address1)
+
+	// the contract account should stop existing at the end of the transaction
+	db.EndTransaction()
+	db.BeginTransaction()
+
+	if db.Exist(address1) {
+		t.Errorf("Account still exists after suicide")
+	}
+}
+
+func TestStateDB_ContractCannotBeCreatedAndDeletedInDifferentTransactions(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	// the account should exist prior to CreateContract call
+	mock.EXPECT().Exists(address1).Return(true, nil)
+	db.CreateContract(address1)
+
+	// start next transaction
+	db.EndTransaction()
+	db.BeginTransaction()
+
+	// suicide the contract from previous transaction
+	db.SuicideNewContract(address1)
+
+	// the contract should still exist when suicided in different transaction
+	db.EndTransaction()
+	db.BeginTransaction()
+
+	if !db.Exist(address1) {
+		t.Errorf("Account should exist after suicide in different transaction")
+	}
+}
+
+func TestStateDB_ContractCreationAndDeletionCanBeRolledBack(t *testing.T) {
```

Comment:
I somehow cannot understand this test.
1. it creates an account
2. it rollbacks via snaphost - i.e. previous did not happen
3. It suicides the contract (which does not exist)
4.  the result is that the contract does exist?

Reply:
In fact it does not have to necessarily create an account. The purpose of `CreateContract` is to mark an account as a contract (eventually creates it if it doesn't exist), so the `Suicide` method knows that it was "created" and can be removed in the same transaction scope.

 The test is verifying, that the `undo` operation on `CreateContract` is done. So in the end the account should still exist, because the `CreateContract` call was rolled back and the `Suicide` method worked as no-op, because there was no preceding `CreateContract` call.

Reply:
I see that the account was not suicided, but how can  it exist? I do not see where it was created at first place?

Reply:
In the mock, I have set, that the state function `Exists` returns true. So when the state DB tries to check, wether the account exists, it gets true from the state and then it stores it into cache. https://github.com/Fantom-foundation/Carmen/blob/0d25cc18cb5920e00ea48ac44390faf3874ecda6/go/state/state_db.go#L485

Reply:
I see now. Perhaps it should be documented in this test as well, the semantics is not that obvious